opt: fold comparison operations during normalization

rytaft opened the pull request:
This PR contains 3 commits:
1. A bug fix for the evaluation of `<tuple> IS (NOT) NULL`
2. Minor fixes to the existing `FoldUnary` and `FoldBinary` rules, as well as the addition of a new `FoldArray` rule
3. Folding of comparison operations with a new `FoldComparison` rule.

See the individual commit messages for details.

rytaft committed:
opt: minor fixes to constant folding rules

This commit fixes a few issues with the implementation
of the FoldBinary and FoldUnary constant folding rules:
1. Some of the tests weren't testing what they were supposed to
   be testing. Fixed the tests and added the expect=<RuleName>
   directive to enforce testing of the rules.
2. The rules were missing opportunities to fold expressions
   involving constant arrays and tuples. Added logic to fold
   these expressions (includes a new rule, FoldArray).
3. The rules were returning all folded constants as Const
   expressions, missing opportunities for further optimization
   of expressions that folded to True, False or Null. Changed
   the custom functions to use the ConstructConstVal function.

Release note: None

rytaft committed:
opt: fold comparison operations during normalization

This commit adds a normalization rule called FoldComparison.
This is one of several rules used to fold constants as part of
the optimizer's normalization phase.

FoldComparison evaluates a comparison operation over constant inputs,
replacing the entire operation with a constant. The rule applies
as long as the evaluation would not cause an error. Any errors should
be saved for execution time, since it's possible that the given
operation will not be executed.

As part of this change, convert the CmpOps slice of builtin functions
to contain *tree.CmpOp rather than tree.CmpOp. This does not cause any
extra allocations, and allows us to avoid copying around these large
objects.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/30022)
<!-- Reviewable:end -->

andy-kimball reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30022)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LM48ScV4ZWIu4QzfJTt:-LM48ScV4ZWIu4QzfJTu:b-clgg2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/49bf65564687a368849d525d64cd91584b1ba68c/pkg/sql/opt/norm/custom_funcs.go#L1074)):*
> ```Go
> func (c *CustomFuncs) IsConst(input memo.GroupID) bool {
> 	ev := memo.MakeNormExprView(c.mem, input)
> 	return ev.IsConstValue() ||
> ```

If a `Tuple` or `Array` only has const children, how come we don't have another rule to turn it into a `Const` with a `DTuple` or `DArray`? I don't love having yet another way to test if something is constant (we already have `(Const)` and `(ConstVal)` - all these different ways will get confused and misused. At the very least, we need a better name for this that doesn't overlap so much.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30022)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/norm/factory.go, line 194 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LM4Ade40QBq0darjTAB:-LM4Ade40QBq0darjTAC:b2iftwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/49bf65564687a368849d525d64cd91584b1ba68c/pkg/sql/opt/norm/factory.go#L194)):*
> ```Go
> // ConstructDatum maps certain datums to separate operators, for easier
> // matching.
> func (f *Factory) ConstructDatum(d tree.Datum) memo.GroupID {
> ```

I just added a `ConstructConstVal` in my prepare PR that we should use instead, since we're constructing more than a `Datum` here.


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
Consider squashing these before you merge so that bisecting still works, unless I misunderstand how you've broken up these commits, but I assume it won't build with just the first.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-:-LM45s9BBdFz8rhpupLD:b-xhnxme)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/memo/typing_test.go, line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LM49i_9950CGJX-VRe_:-LM49i_9950CGJX-VRea:b-srodd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/df08a3a2f814f18de29ba196c0318ea21586c6df/pkg/sql/opt/memo/typing_test.go#L142)):*
> ```Go
> // TestTypingComparisonAssumptions ensures that comparison overloads conform to
> // certain assumptions we're making in the type inference code:
> //   1. The overload can be inferred from the operator type and the data
> ```

nit: kind of odd to number this if there's only one? I like this test a lot though!
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LM48ScV4ZWIu4QzfJTt:-LM4CIiR8NglDdd5YxWz:b6ab9ej) ([raw file](https://github.com/cockroachdb/cockroach/blob/49bf65564687a368849d525d64cd91584b1ba68c/pkg/sql/opt/norm/custom_funcs.go#L1074)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

If a `Tuple` or `Array` only has const children, how come we don't have another rule to turn it into a `Const` with a `DTuple` or `DArray`? I don't love having yet another way to test if something is constant (we already have `(Const)` and `(ConstVal)` - all these different ways will get confused and misused. At the very least, we need a better name for this that doesn't overlap so much.
</blockquote></details>

Won't this complicate span generation for tuples? Then we have a different situation for `(a, b, c) = (1, 2,3)` vs. `(a, b, c) = (1, 2, d)`?


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30022)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LM48ScV4ZWIu4QzfJTt:-LM4Lx-c2aFph9MUpbYm:b-n1sm74) ([raw file](https://github.com/cockroachdb/cockroach/blob/49bf65564687a368849d525d64cd91584b1ba68c/pkg/sql/opt/norm/custom_funcs.go#L1074)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Won't this complicate span generation for tuples? Then we have a different situation for `(a, b, c) = (1, 2,3)` vs. `(a, b, c) = (1, 2, d)`?
</blockquote></details>

Yes, that's true; pattern recognition becomes tougher. This is a larger question, though. For example, take the `FoldNonNullIsNull` rule. Shouldn't that use this new definition of what "constant" is? I don't think we should create a new definition without reconciling it with the old definition, even though that means more work.

I'd be OK with not making a new definition (and just not being able to fold `Tuple`/`Array` cases for now), or with making a new reconciled definition (and changing the various places that are affected). But I don't think we should have two side-by-side definitions that aren't consistent.


<!-- Sent from Reviewable.io -->

rytaft committed:
sql: fix evaluation of <tuple> IS NULL

Prior to this patch, the comparison op overload for the
IsNotDistinctFrom operation with tuple inputs had NullableArgs
incorrectly set to false. As a result, the output was incorrect
for queries such as:

SELECT (1, 2, 3) IS NULL

This query should return false, but instead it returned NULL.
This commit fixes this bug.

Release note (bug fix): Fixed the evaluation of <tuple> IS NOT NULL and
<tuple> IS NULL comparison operations involving a non-null constant tuple
to return true or false rather than NULL.

rytaft reviewed (commented):
I've fixed a few bugs and made some updates. PTAL.

> Consider squashing these before you merge so that bisecting still works, unless I misunderstand how you've broken up these commits, but I assume it won't build with just the first.

@justinj I think it should build with just the first (now the second) commit... was there something that made you think it wouldn't?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-:-LM9L-MIBa04_QJ9RdDQ:b-owhtuw)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/memo/typing_test.go, line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LM49i_9950CGJX-VRe_:-LM4NYhC7jlC148MzJ4j:b-epdsxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/df08a3a2f814f18de29ba196c0318ea21586c6df/pkg/sql/opt/memo/typing_test.go#L142)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: kind of odd to number this if there's only one? I like this test a lot though!
</blockquote></details>

I agree it's a bit odd in isolation, but I was just following the convention in the rest of the file... `TestTypingUnaryAssumptions` also has just one, so it seemed better to just follow that convention
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LM48ScV4ZWIu4QzfJTt:-LM9JmKs1epJinGDbtYs:bq5mg5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/49bf65564687a368849d525d64cd91584b1ba68c/pkg/sql/opt/norm/custom_funcs.go#L1074)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Yes, that's true; pattern recognition becomes tougher. This is a larger question, though. For example, take the `FoldNonNullIsNull` rule. Shouldn't that use this new definition of what "constant" is? I don't think we should create a new definition without reconciling it with the old definition, even though that means more work.

I'd be OK with not making a new definition (and just not being able to fold `Tuple`/`Array` cases for now), or with making a new reconciled definition (and changing the various places that are affected). But I don't think we should have two side-by-side definitions that aren't consistent.
</blockquote></details>

I added a rule to fold arrays, but tuples are complicated because we use them in several different contexts -- for span generation as Justin mentioned, and also in VALUES clauses, etc. 

So I changed the name of this function to `IsConstValueOrTuple`, and updated the other rules that use `ConstValue` to use this new function instead. Does this seem like a good solution?

I don't really want to exclude tuples from constant folding, because then we'll miss out on common comparison expressions like `SELECT 1 IN (1, 2, 3)`
___
*[pkg/sql/opt/norm/factory.go, line 194 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LM4Ade40QBq0darjTAB:-LM4NR9M5PSNrUlcESVR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/49bf65564687a368849d525d64cd91584b1ba68c/pkg/sql/opt/norm/factory.go#L194)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I just added a `ConstructConstVal` in my prepare PR that we should use instead, since we're constructing more than a `Datum` here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, good work

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-:-LMEOotBD8YB0RIXyklF:bxo1df3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt/memo/typing.go, line 368 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LMENMAVCK_nuBQ44C41:-LMENMAVCK_nuBQ44C42:b9ebzsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/30e18e5dd38f0ed114d2844fb6a2a1bcea1599c0/pkg/sql/opt/memo/typing.go#L368)):*
> ```Go
> 	// TestTypingComparisonAssumptions test ensures this will be the case even if
> 	// new operators or overloads are added.
> 	for _, binOverloads := range tree.CmpOps[compOp] {
> ```

NIT: binOverloads => cmpOverloads
___
*[pkg/sql/opt/norm/testdata/rules/fold_constants, line 20 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LMEOYeu4c6GD5RD3kNf:-LMEOYeu4c6GD5RD3kNg:b-9987nr) ([raw file](https://github.com/cockroachdb/cockroach/blob/30e18e5dd38f0ed114d2844fb6a2a1bcea1599c0/pkg/sql/opt/norm/testdata/rules/fold_constants#L20)):*
> ```
>       └── const: ARRAY[1,2,3] [type=int[]]
> 
> opt expect=FoldArray
> ```

Also add an interesting negative case, like an array that only has constants except for the last element.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Thanks!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-:-LMEak63EX6F1dmdOc6N:b-voiyyw)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt/memo/typing.go, line 368 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LMENMAVCK_nuBQ44C41:-LME_Y_G4RXDyD_bOElp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30e18e5dd38f0ed114d2844fb6a2a1bcea1599c0/pkg/sql/opt/memo/typing.go#L368)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: binOverloads => cmpOverloads
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/testdata/rules/fold_constants, line 20 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/30022#-LMEOYeu4c6GD5RD3kNf:-LME_fS4EFun1CL07Ue1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30e18e5dd38f0ed114d2844fb6a2a1bcea1599c0/pkg/sql/opt/norm/testdata/rules/fold_constants#L20)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Also add an interesting negative case, like an array that only has constants except for the last element.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft commented:
bors r-





<!-- Sent from Reviewable.io -->

rytaft commented:
bors r+





<!-- Sent from Reviewable.io -->
